Metric Configuration Store: Refactor & Connect UI Components to the Store (2/2)

mxosman opened the pull request:
## Description of the change

Equip with a new mobx data store, this PR refactors and connects all Metric Configuration & related UI components to the store. This should be a no-op refactor with regards to behavior (everything here should behave exactly the same as it does in staging). 

## Related issues

Closes #134 

## Checklists

### Development

**This box MUST be checked by the submitter prior to merging**:
- [x] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

mxosman commented on publisher/src/stores/MetricConfigStore.tsx:
```diff
@@ -152,93 +216,121 @@ class MetricConfigStore {
-    });
-  }
+    } catch (error) {
+      return new Error(error as string);
```
Added a try/catch here since `getMetricSettings` can throw an error - this will catch it and return it as an Error object to the UI layer to be handled.

mxosman commented on publisher/src/stores/MetricConfigStore.tsx:
```diff
@@ -152,93 +216,121 @@ class MetricConfigStore {
+      this.metrics[systemMetricKey].description = metadata.description;
+      this.metrics[systemMetricKey].frequency =
+        metadata.frequency as ReportFrequency;
+    }
```
This is where we add the necessary metadata required for rendering. During the store's initialization, an optional `metadata` object is passed in and this method loads those details on the metric object we use to update and render from.

mxosman commented on publisher/src/stores/MetricConfigStore.tsx:
```diff
@@ -337,6 +460,38 @@ class MetricConfigStore {
+      if (isDisaggregationDisabledAndOneDimensionReEnabled) {
+        this.disaggregations[systemMetricKey][disaggregationKey].enabled = true;
+      }
+    }
```
Special handling for when last dimension is turned off and when everything in a breakdown is turned off and you're turning back on one dimension.

mxosman commented on publisher/src/stores/MetricConfigStore.tsx:
```diff
@@ -298,6 +400,26 @@ class MetricConfigStore {
+          ].enabled = enabledStatus;
+        }
+      );
+    }
```
Special handling for when a disaggregation is disabled/enabled.

mxosman commented on publisher/src/stores/MetricConfigStore.tsx:
```diff
@@ -102,16 +119,59 @@ class MetricConfigStore {
+
+      return metrics;
+    }
+  };
```
This felt useful to have to render a list of metrics by system.

mxosman commented on publisher/src/pages/Settings.tsx:
```diff
@@ -46,31 +46,17 @@ const Settings = () => {
-            setActiveMetricKey={setActiveMetricKey}
-            setListOfMetrics={setListOfMetrics}
-          />
-        )}
```
🧹 🧹 🧹

mxosman commented on publisher/src/components/MetricConfiguration/MetricConfiguration.tsx:
```diff
@@ -52,413 +48,71 @@ import {
-
-      return prev;
-    });
-  };
```
BYE! 👋 👋 👋

mxosman commented on publisher/src/components/MetricConfiguration/MetricConfiguration.tsx:
```diff
@@ -52,413 +48,71 @@ import {
-    } else {
-      saveMetricSettings(updatedSetting);
-    }
-  };
```
So, we don't need both save functions - the only component so far that has ever used the debouncing is the ContextConfiguration (text field context), so it made sense to just have the `debouncedSave` function live there and clean up the save function.

mxosman commented on publisher/src/components/MetricConfiguration/Configuration.tsx:
```diff
@@ -47,222 +44,247 @@ import {
-        );
+    const activeDisaggregationKeys =
+      disaggregations[systemMetricKey] &&
+      Object.keys(disaggregations[systemMetricKey]);
```
Not sure if this was wise, but I decided I just needed an array of keys and from there, with our structure (`{ [key: string]: obj }`), I have quick access to the key for components that need it and can quickly look up the object I need to get the rest of the details.

mxosman replied:
Could've used `Object.entries` and used that multi-dimensional array, but this felt simple enough. No strong feelings!

mxosman commented on publisher/src/components/MetricConfiguration/Configuration.tsx:
```diff
@@ -47,222 +44,247 @@ import {
+      },
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      [systemMetricKey]
+    );
```
Whenever a new metric is selected, reset the disaggregation tabs so the first one is open.

terryttsai commented on publisher/src/components/MetricConfiguration/MetricConfiguration.tsx:
```diff
@@ -471,14 +125,16 @@ export const MetricConfiguration: React.FC<{
-        async (currentAgencyId, previousAgencyId) => {
+        async (_, previousAgencyId) => {
           if (previousAgencyId !== undefined) {
             setIsLoading(true);
```
Perhaps the body of the function here can be reused in the previous useEffect? Basically can we create one function that initializes the store and sets the active system and metric?

mxosman replied:
Great call!

terryttsai commented on publisher/src/components/MetricConfiguration/ContextConfiguration.tsx:
```diff
@@ -32,189 +33,185 @@ import {
+                      checked={currentContext.value === "yes"}
+                      onChange={() => {
+                        const updatedSetting = updateContextValue(
+                          activeSystem as string,
```
Instead of "as string", can the argument type just be defined as MetricSystem?

mxosman replied:
Good question - activeSystem could technically be `undefined` (before the initialization step) - though the initialization happens and it is defined as the first system before render this component. I'll have to rework the function a bit to handle the undefined - if I specify AgencySystems, it'll know it can also be undefined and if I include undefined in these functions it throws a cascade of TS errors. Maybe all the updater functions can short circuit in the off chance undefined passes through? Any suggestions?

terryttsai commented on publisher/src/components/MetricConfiguration/ContextConfiguration.tsx:
```diff
@@ -32,189 +33,185 @@ import {
+                      onChange={() => {
+                        const updatedSetting = updateContextValue(
+                          activeSystem as string,
+                          activeMetricKey as string,
```
can this be undefined?

terryttsai replied:
Maybe the function can just throw an error or log a warning if undefined?

mxosman replied:
Yeah - great call - this is tricky. Technically they can, but things are initialized by the time it gets to this point. 

Let me think about this - if we update the function to throw an error if undefined, then we'll need an extra step of checking if `updatedSetting` is undefined before saving, right?

mxosman replied:
Worked it out in this commit: [ed985e5](https://github.com/Recidiviz/justice-counts/pull/132/commits/ed985e5bd31081d7c07409f72567d7cd4af03033)

Thank you!

terryttsai commented on publisher/src/components/MetricConfiguration/Configuration.tsx:
```diff
@@ -47,222 +44,248 @@ import {
-        filteredMetricSettings[activeMetricKey]?.disaggregations?.find(
-          (disaggregation) => disaggregation.key === activeDisaggregation.key
-        );
+    const activeDimensionKeys = (activeDisaggregationKey &&
```
Could write this as
```
    const activeDimensionKeys = (activeDisaggregationKey &&
      dimensions[systemMetricKey]?.[activeDisaggregationKey] ?
      Object.keys(
        dimensions[systemMetricKey][activeDisaggregationKey]
      ) : []);
```

and then you don't need to cast types

mxosman replied:
Nice!

terryttsai commented on publisher/src/components/MetricConfiguration/MetricDefinitions.tsx:
```diff
@@ -37,200 +33,255 @@ import {
-  ) => void;
+  activeDimensionKey: string | undefined;
+  activeDisaggregationKey: string | undefined;
+  saveUpdatedMetricSettings: (updatedSetting: MetricSettings) => void;
```
I wonder if it's worth making the store handle showing the error toast, then we wouldn't need to pass this in as a prop?

mxosman replied:
Yeah - I was back and forth on this. Error toast felt more of a responsibility of the UI layer - but if it's in the store, then we get the benefit of not having to prop drill this. Let me try showing the toast in the store - that might work out nicer.

mxosman replied:
Feels so much nicer centralizing the save function to the store! Thank you for all of these fantastic suggestions!

terryttsai commented on publisher/src/components/MetricConfiguration/MetricDefinitions.tsx:
```diff
@@ -37,200 +33,255 @@ import {
+      ? metricDefinitionSettingsKeys
+      : dimensionDefinitionSettingsKeys;
+
+    const selectionOptions: MetricConfigurationSettingsOptions[] = [
```
Can we define MetricConfigurationSettingsOptions as 
```
export const MetricConfigurationSettingsOptions = ["Yes", "No", "N/A"] as const;
export type MetricConfigurationSettingsOptions = typeof MetricConfigurationSettingsOptions;
```

so we have both the array of strings as well as the type?

mxosman replied:
That's a good idea - I was just using the one in the `common` repo and wanted to avoid having to update that repo in this PR if it wasn't necessary.

I tried what you implemented above and it turned the type into a readonly array - wasn't sure how to shake this TS warning that came as a result:
<img width="883" alt="Screen Shot 2022-11-02 at 6 40 17 PM" src="https://user-images.githubusercontent.com/59492998/199624245-c9100965-6386-4063-b875-86c661b0cd1b.png">

Any ideas?

mxosman replied:
Let me think about this a little more - I might have an idea of what's going on.

mxosman replied:
Hmm. I do like this pattern and have used it in the past in the app, but as I keep exploring it - it feels like more trouble than its worth. This required updating everything that's typed as `MetricConfigurationSettingsOptions` and was running into the same compatibility issues as above.

mxosman replied:
OK! I think I figured out what was going on - it needs to be used like this to work as intended:

```
export const metricConfigurationSettingsOptions = ["Yes", "No", "N/A"] as const;
export type MetricConfigurationSettingsOptions =
  typeof metricConfigurationSettingsOptions[number];
```

mxosman replied:
[b9418ff](https://github.com/Recidiviz/justice-counts/pull/132/commits/b9418ffdd8a95f4bc7859874a4972ce7ca6f52e7)

Thanks for this optimization! This is cleaning up very nicely!

terryttsai replied:
Thanks for looking into this!

mxosman commented on publisher/src/components/MetricConfiguration/MetricConfiguration.tsx:
```diff
@@ -52,413 +48,71 @@ import {
-  | "METRIC_SETTING"
-  | "DIMENSION_SETTING";
-
 export type MetricSettings = {
```
It could!

terryttsai commented on publisher/src/components/MetricConfiguration/Configuration.tsx:
```diff
@@ -123,11 +123,11 @@ export const Configuration: React.FC<MetricConfigurationProps> = observer(
               onChange={() => {
                 const updatedSetting = updateMetricEnabledStatus(
-                  activeSystem as string,
+                  activeSystem,
```
I see, so originally we were doing type casting here because we're only using this if an activeSystem and activeMetricKey are set in the store but the types could be undefined in the store. 

I guess we'll be moving away from setting activeSystem and activeMetricKey in the store anyways so actually I'd be cool with just wrapping this in a conditional like this:

```
              onChange={() => {
                if (activeSystem && activeMetricKey) {
                  const updatedSetting = updateMetricEnabledStatus(
                    activeSystem,
                    activeMetricKey,
                    true
                  );
                  saveMetricSettings(updatedSetting);
                }
              }}
```

Or we could keep the type casting since we expect them to be strings? Sorry I didn't give much thought to this yesterday!

terryttsai replied:
My preference is to only call this `updateMetricEnabledStatus` if we have the required information otherwise don't call it

mxosman replied:
Ah, OK. Makes sense - let me go back and revert all of those changes.

terryttsai reviewed (approved):
Thanks for addressing all the feedback! Left a small comment but everything looks good!

mxosman commented:
> Thanks for addressing all the feedback! Left a small comment but everything looks good!

Thank you for all of your help with this - this already feels so much better to interface with!
